components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.core.internal;
+
+import org.apache.axis2.engine.AxisConfiguration;
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.lang.StringUtils;
+import org.wso2.carbon.base.ServerConfiguration;
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.identity.core.ServiceURLBuilder;
+import org.wso2.carbon.identity.core.URLBuilderException;
+import org.wso2.carbon.identity.core.model.ServiceURL;
+import org.wso2.carbon.identity.core.util.IdentityCoreConstants;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.utils.CarbonUtils;
+import org.wso2.carbon.utils.NetworkUtils;
+
+import java.net.SocketException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.core.util.IdentityTenantUtil.isTenantQualifiedUrlsEnabled;
+
+/**
+ * Implementation for {@link ServiceURLBuilder}.
+ * Builder for {@link ServiceURL} instances.
+ */
+public class DefaultServiceURLBuilder implements ServiceURLBuilder {
+
+    private String[] urlPaths;
+    private Map<String, String> parameters;
+    private String fragment;
+    private Map<String, String> fragmentParams;
+
+    /**
+     * This method is called to add the URL path to the builder.
+     *
+     * @param paths Context path. Can provide multiple context paths with a comma separated string.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addPath(String ... paths) {
+
+        this.urlPaths = paths;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with other parameters. Such parameters should be
+     * entered as <k,v> pairs.
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addParameter(String key, String value) {
+
+        parameters = new HashMap<>();
```

Comment:
Why create a new hashmap here. It should be pre-initialized.
This avoids invoking this method multiple times and adding several parameters